mobile/src/main/java/ch/epfl/sdp/mobile/ui/speech_recognition/ChessSpeechRecognitionFilter.kt

```diff
@@ -0,0 +1,210 @@
+package ch.epfl.sdp.mobile.ui.speech_recognition
+
+typealias Word = String // Piece | Letter | Number
+
+typealias Speech = String // Raw speech Sentence
+
+typealias ChessPiece = String
+
+typealias ChessNumber = String
+
+typealias ChessLetter = String
+
+typealias ChessWord = String // Filter Result
+
+/**
+ * A Chess speech recognition results filters that factors a chess command given a vocabulary
+ * dictionary and a set of filtering rules
+ *
+ * @property chessDictionary chess dictionary rules
+ * @property speechRules chess filtering rules
+ */
+class ChessSpeechRecognitionFilter(
+    private val chessDictionary: ChessDictionary = ChessSpeechDictionary,
+    private val speechRules: ChessSpeechFilterRules = ChessSpeechFilterRules
+) : SpeechRecognitionFilter {
+
+  /**
+   * Applies all speechRules on speeches
+   * @param speeches list of tokenized speeches
+   * @param confidencesScores list of confidence score for recognized speech
+   * @return chess keyword string after applying all speechRules, or an empty string if all rules
+   * fail
+   */
+  private fun applyRules(
+      speeches: List<List<Word>>,
+      confidencesScores: List<Double>,
+  ): String {
+
+    val candidates = mutableMapOf<String, Double>()
+    for (chessRule in speechRules.rulesSet) {
+      speeches.forEachIndexed { index, speech ->
+        if (chessRule.rule(speech)) {
+          candidates[chessRule.chessPiece] = confidencesScores[index]
+        }
+      }
+    }
+    return bestCandidate(candidates)
+  }
+
+  /**
+   * Returns detected chess keyword candidate with best confidence score
+   * @param candidates mapping from detected in dictionary chess keywords to their scores
+   * @return candidate as a chess keyword
+   */
+  private fun bestCandidate(candidates: Map<Word, Double>): String {
+    if (candidates.isEmpty()) {
+      return ""
+    }
+
+    if (candidates.size == 1) {
+      return candidates.entries.first().key
+    }
```

Comment:
I'm not sure we need to special-case this, since the for loop with a single entry seems to have the same result.

Reply:
Yes indeed !